### PR TITLE
feat: implement off-chain-auth solution

### PR DESCRIPTION
**Pull request by ruojunm**
### Description

For those dapp users, we here provide an off-chain-auth solution to access SP apis, so that  they don't have to make the signatures via the wallet popups over again and again.

### Changes

Notable changes: 
* add OffChainAuthKeyTable to store user's EDDSA public key
* add "request nonce" API to return the current nonce to the clients so that they can generate EDDSA key pairs based on the given nonce
* add "update user key" API to store user's EDDSA public key, so that later we can verify user's EDDSA sig 
* add "verify off chain auth sig" API and apply it in the request_uilt.go 's  verifySignature method
* add "verifyPersonalSignature" method in request_uilt.go, so that dapp users can use the wallet to sign the request to update their EDDSA public key
* add unit test

### CODE REVIEW

File: docs/run-book/01-deployment.md [hunk 1]
```diff
@@ -59,6 +59,7 @@ Build   : go1.20.1 darwin amd64 2023-03-04 23:54
 ```shell
 # show services list
 ./gnfd-sp list
+auth                 off chain auth server
```
Comment (will-2012) [inline docs/run-book/01-deployment.md:62]: Best to keep a consistent style, verbs that start with.

File: model/const.go [hunk 2]
```diff
@@ -160,6 +165,10 @@ const (
 	// SignTypeV2 is an authentication algorithm, which is used by metamask
 	SignTypeV2 = "authTypeV2"
 
+	SignTypeOffChain   = "OffChainAuth" // sign type - off-chain-auth
```
Comment (will-2012) [inline model/const.go:168]: When to remove SignTypeV2?

File: store/mock/mockSPDB.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
+package mock
```
Comment (will-2012) [inline store/mock/mockSPDB.go:1]: `mockgen -source=store/sqldb/database.go -destination=store/sqldb/database_mock.go -package=sqldb` has existed in the makefile.

File: model/const.go (hunk 2 above)
Comment (ruojunm) [reply model/const.go:170]: After Dcellar uses off-chain-auth, I will remove SignTypeV2.  So far, let's keep it and leave some time for Dcellar

**Review by constwz — APPROVED** [review]
LGTM

File: model/const.go [hunk 4]
```diff
@@ -42,6 +44,7 @@ var SpServiceDesc = map[string]string{
 	MetadataService:    "Provides the ability to query meta data",
 	BlockSyncerService: "Syncs block data to db",
 	P2PService:         "Communicates with SPs on p2p protocol",
+	AuthService:        "off chain auth server",
```
Comment (annielz) [inline model/const.go:47]: Description start with upper case

**Review by 2129zxl — APPROVED** [review]
LGTM

**Review by krish-nr — APPROVED** [review]
LGTM

**Comment by BarryTong65** [comment]
LGTM

**Review by BarryTong65 — APPROVED** [review]

File: model/errors/rpc_error.go [hunk 5]
```diff
@@ -98,6 +98,15 @@ var (
 	ErrInvalidAccountID = errors.New("invalid account id")
 )
 
+// auth service error
+var (
+	ErrInvalidRegNonce           = errors.New("invalid regNonce")
```
Comment (actioncli) [inline model/errors/rpc_error.go:103]: add error description?

**Review by annielz — APPROVED** [review]

File: model/errors/rpc_error.go (hunk 5 above)
Comment (ruojunm) [reply model/errors/rpc_error.go]: Good catch. They are actually not used. I deleted all of them.

File: model/const.go (hunk 4 above)
Comment (ruojunm) [reply model/const.go]: Good Catch. Fixed.

**Review by joeylichang — APPROVED** [review]

**Pull request merged**